lib/internal/test_runner/utils.js

```diff
@@ -74,10 +87,72 @@ function convertStringToRegExp(str, name) {
   }
 }
 
+const kBuiltinDestinations = new SafeMap([
+  ['stdout', process.stdout],
+  ['stderr', process.stderr],
+]);
+
+const kBuiltinReporters = new SafeMap([
+  ['spec', 'node:test/reporter/spec'],
+  ['dot', 'node:test/reporter/dot'],
+  ['tap', 'node:test/reporter/tap'],
+]);
+
+const kDefaultReporter = 'tap';
+const kDefaltDestination = 'stdout';
+
+async function getReportersMap(reporters, destinations) {
+  const result = await SafePromiseAll(ArrayPrototypeMap(reporters, async (name, i) => {
+    const destination = kBuiltinDestinations.get(destinations[i]) ?? createWriteStream(destinations[i]);
+    let reporter = await requireOrImport(kBuiltinReporters.get(name) ?? name);
+
+    if (reporter && reporter.default) {
+      reporter = reporter.default;
+    }
+
+    if (reporter.prototype && ObjectGetOwnPropertyDescriptor(reporter.prototype, 'constructor')) {
+      reporter = new reporter();
+    }
+
+    if (!reporter) {
+      throw new ERR_INVALID_ARG_VALUE('Reporter', name, 'is not a valid reporter');
+    }
+
+    return { __proto__: null, reporter, destination };
+  }
+  ));
+  return result;
+}
+
+
+async function setupTestReporters(testsStream) {
+  const destinations = getOptionValue('--test-reporter-destination');
+  const reporters = getOptionValue('--test-reporter');
+
+  if (reporters.length === 0 && destinations.length === 0) {
+    ArrayPrototypePush(reporters, kDefaultReporter);
+  }
+
+  if (reporters.length === 1 && destinations.length === 0) {
+    ArrayPrototypePush(destinations, kDefaltDestination);
+  }
+
+  if (destinations.length !== reporters.length) {
+    throw new ERR_INVALID_ARG_VALUE('--test-reporter', reporters,
+                                    'must match the number of specified \'--test-reporter-destination\'');
+  }
+
+  const reportersMap = await getReportersMap(reporters, destinations);
+  for (const { reporter, destination } of reportersMap) {
```

Comment:
does SafePromiseAll return an array whose `Symbol.iterator` is also safe? if not, `for-of` should be avoided here.

Reply:
The "safe" thing about `SafePromiseAll` is that:

- it takes an array and pass it though `SafeArrayIterator` before passing it to the built-in.
- it ensures intermediates promises are created from the `SafePromise` class so there are not subject to `%Promise.prototype%.then` mutation.

However, it does return a classique `Array`, so nothing's inherently safe about its return value.